### ISSUE TITLE
Increased error rate: Transport endpoint is not connected (os error 107)

### ISSUE THREAD

**Issue by kforsthoevel** [linkerd/linkerd2#5520]
### What is the issue?

Since the upgrade from `stable-2.9.1` to `edge-20.12.4` we see an increased error rate (~200/min) in the `linkerd-proxy`. The high rate of errors that haven't been there before worries me a lot.

### Logs, error output, etc

```json
{
  "timestamp": "[332714.492430s]",
  "level": "INFO",
  "fields": {
    "message": "Connection closed",
    "error": "Transport endpoint is not connected (os error 107)"
  },
  "target": "linkerd2_app_core::serve",
  "span": {
    "field_error": "expected value at line 1 column 1",
    "name": "accept"
  },
  "spans": [
    {
      "field_error": "EOF while parsing a value at line 1 column 0",
      "name": "inbound"
    },
    {
      "field_error": "expected value at line 1 column 1",
      "name": "accept"
    }
  ],
  "threadId": "ThreadId(1)"
}
```

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
√ cli is up-to-date

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 20.12.4 but the latest edge version is 21.1.1
    see https://linkerd.io/checks/#l5d-version-control for hints

linkerd-ha-checks
-----------------
√ pod injection disabled on kube-system
√ multiple replicas of control plane pods

linkerd-prometheus
------------------
√ prometheus add-on service account exists
√ prometheus add-on config map exists
√ prometheus pod is running

linkerd-grafana
---------------
√ grafana add-on service account exists
√ grafana add-on config map exists
√ grafana pod is running

Status check results are √

```

### Environment

- Kubernetes Version: v1.18.14
- Cluster Environment: kops 
- Host OS: Ubuntu 20.04 LTS
- Linkerd version: edge-20.12.4

Does anyone knows what this error means and what we can do to get rid of the error? Thanks.

**Comment by hawkw** [comment]
The `field_error` messages in 
```json
  "span": {
    "field_error": "expected value at line 1 column 1",
    "name": "accept"
  },
  "spans": [
    {
      "field_error": "EOF while parsing a value at line 1 column 0",
      "name": "inbound"
    },
    {
      "field_error": "expected value at line 1 column 1",
      "name": "accept"
    }
  ],
```
are not a proxy error, but [an error emitted by the JSON log formatter](https://github.com/tokio-rs/tracing/blob/b1baa6c2ef4877bd7f11de5991234fc81a025947/tracing-subscriber/src/fmt/format/json.rs#L165-L168), which likely signifies a bug in the JSON formatter that was encountered while trying to log the "connection refused" error.

The actual error being logged by the proxy is this part:
```json
  "fields": {
    "message": "Connection closed",
    "error": "Transport endpoint is not connected (os error 107)"
  },
```

**Comment by adleong** [comment]
@hawkw is there an issue for the bug in the tokio tracing json formatting that we can link to?

**Comment by olix0r** [comment]
@kforsthoevel are these connection closed messages correlated to request failures or other application errors? This message typically indicates that the peer has closed the connection, which may be benign. Specifically, there were caching changes made in December, and these message may simply indicate we're closing idle connections more aggressively.

**Comment by kforsthoevel** [comment]
@olix0r Sorry, we have switched to stable-2.9.2 this morning and all those log messages disappeared. We have had around 3000 logs messages from all `linkerd-proxy` containers per 10 minutes and now we have around 100 per 10 minutes w/o changing anything else.
We weren't aware of any actual problems, but the sheer increase in log messages got me worried a lot.

**Comment by olix0r** [comment]
@kforsthoevel Thanks for the clarification. For what it's worth, stable-2.9.2 does **not** include a proxy update, so this issue likely persists on edge releases...

At the very least, we need to fix the log message formatting; but we should also investigate the volume of connection resets... We could consider moving these log messages to the DEBUG level to avoid excessive logging, though I'm not sure if that's the best course of action, as this may be a symptom of another legitimate issue.

**Comment by hawkw** [comment]
> @hawkw is there an issue for the bug in the tokio tracing json formatting that we can link to?

opened an upstream issue: tokio-rs/tracing#1192

**Comment by hawkw** [comment]
> > @hawkw is there an issue for the bug in the tokio tracing json formatting that we can link to?
> 
> opened an upstream issue: [tokio-rs/tracing#1192](https://github.com/tokio-rs/tracing/issues/1192)

Update: turns out that was a proxy issue, not an upstream bug --- we weren't configuring the logger correctly. Fixed in linkerd/linkerd2-proxy#845.

**Comment by adleong** [comment]
Fixed by linkerd/linkerd2-proxy#845
